// @apollo/federation-subgraph-compatibility-tests/dist/tests/inaccessible.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const client_1 = require("../utils/client");
const graphql_1 = require("graphql");
describe('@inaccessible', () => {
    it('should return @inaccessible directives in _service sdl', async () => {
        const response = await (0, client_1.productsRequest)({
            query: 'query { _service { sdl } }',
        });
        const { sdl } = response.data._service;
        const normalizedSDL = (0, graphql_1.stripIgnoredCharacters)(sdl);
        expect(normalizedSDL).not.toContain('@federation__inaccessible');
        expect(normalizedSDL).toContain('unit:String@inaccessible');
    });
    it('should be able to query @inaccessible fields via the products schema directly', async () => {
        const resp = await (0, client_1.productsRequest)({
            query: `
        query GetProduct($id: ID!) {
          product(id: $id) {
            dimensions {
              unit
            }
          }
        }
      `,
            variables: { id: 'apollo-federation' },
        });
        expect(resp).not.toHaveProperty('errors');
        expect(resp).toMatchObject({
            data: {
                product: {
                    dimensions: {
                        unit: 'kg',
                    },
                },
            },
        });
    });
});
