// @apollo/federation-subgraph-compatibility-tests/dist/tests/override.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const client_1 = require("../utils/client");
const graphql_1 = require("graphql");
describe('@override', () => {
    it('should return @override directives in _service sdl', async () => {
        const response = await (0, client_1.productsRequest)({
            query: 'query { _service { sdl } }',
        });
        const { sdl } = response.data._service;
        expect((0, graphql_1.stripIgnoredCharacters)(sdl)).toMatch(/(@override|@federation__override)\(from:"users"\)/);
    });
    it('should return overridden user name', async () => {
        const resp = await (0, client_1.routerRequest)({
            query: `
        query GetProduct($id: ID!) {
          product(id: $id) {
            createdBy {
              name
            }
          }
        }
      `,
            variables: { id: 'apollo-federation' },
        });
        expect(resp).not.toHaveProperty('errors');
        expect(resp).toMatchObject({
            data: {
                product: {
                    createdBy: {
                        name: 'Jane Smith',
                    },
                },
            },
        });
    });
});
