// @apollo/federation-subgraph-compatibility-tests/dist/tests/provides.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const client_1 = require("../utils/client");
describe('@provides', () => {
    test('should return @provides directives in _service sdl', async () => {
        const response = await (0, client_1.productsRequest)({
            query: 'query { _service { sdl } }',
        });
        const { sdl } = response.data._service;
        expect((0, graphql_1.stripIgnoredCharacters)(sdl)).toMatch(/createdBy:User(@provides|@federation__provides)\(fields:"totalProductsCreated"\)/);
    });
    test('should return locally computed @provides field', async () => {
        const resp = await (0, client_1.productsRequest)({
            query: `#graphql
      query ($id: ID!) {
        product(id: $id) {
          createdBy { email totalProductsCreated }
        }
      }`,
            variables: { id: 'apollo-federation' },
        });
        expect(resp).not.toHaveProperty('errors');
        expect(resp).toMatchObject({
            data: {
                product: {
                    createdBy: {
                        email: 'support@apollographql.com',
                        totalProductsCreated: expect.any(Number),
                    },
                },
            },
        });
        const totalProductsCreated = resp.data.product.createdBy.totalProductsCreated;
        expect(totalProductsCreated).not.toEqual(4);
    });
});
