// @apollo/federation-subgraph-compatibility-tests/dist/tests/requires.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const client_1 = require("../utils/client");
describe('@requires', () => {
    test('should return @requires directives in _service sdl', async () => {
        const response = await (0, client_1.productsRequest)({
            query: 'query { _service { sdl } }',
        });
        const { sdl } = response.data._service;
        expect((0, graphql_1.stripIgnoredCharacters)(sdl)).toMatch(/averageProductsCreatedPerYear:Int(@requires|@federation__requires)\(fields:"totalProductsCreated yearsOfEmployment"\)/);
    });
    test('should return field with computed values from @requires field set', async () => {
        const resp = await (0, client_1.routerRequest)({
            query: `#graphql
        query ($id: ID!) {
          product(id: $id) { createdBy { averageProductsCreatedPerYear email } }
        }`,
            variables: { id: 'apollo-federation' },
        });
        expect(resp).not.toHaveProperty('errors');
        expect(resp).toMatchObject({
            data: {
                product: {
                    createdBy: {
                        averageProductsCreatedPerYear: expect.any(Number),
                        email: 'support@apollographql.com',
                    },
                },
            },
        });
    });
});
