// @apollo/federation-subgraph-compatibility-tests/dist/tests/shareable.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const client_1 = require("../utils/client");
const graphql_1 = require("graphql");
describe('@shareable', () => {
    it('should return @shareable directives in _service sdl', async () => {
        const response = await (0, client_1.productsRequest)({
            query: 'query { _service { sdl } }',
        });
        const { sdl } = response.data._service;
        expect((0, graphql_1.stripIgnoredCharacters)(sdl)).toMatch(/type ProductDimension(@shareable|@federation__shareable)/);
    });
    it('should be able to resolve @shareable ProductDimension types', async () => {
        const resp = await (0, client_1.routerRequest)({
            query: `
        query GetProduct($id: ID!) {
          product(id: $id) {
            dimensions {
              size
              weight
            }
          }
        }
      `,
            variables: { id: 'apollo-federation' },
        });
        expect(resp).not.toHaveProperty('errors');
        expect(resp).toMatchObject({
            data: {
                product: {
                    dimensions: {
                        size: 'small',
                        weight: 1,
                    },
                },
            },
        });
    });
});
